README.md
# STM32 Servo Library using HardwareTimer from Arduino_Core_STM32

PWM signal for Servo motor control is generated using HardwareTimer 
library from Arduino_Core_STM32, for every i/o pin, that has PWM 
capability. Timer and channel are configured within the attach() method.

## Requirements
[Arduino](https://www.arduino.cc/)<br>
[Arduino_Core_STM32](https://github.com/stm32duino/Arduino_Core_STM32)<br>

This library is tested with
**Arduino IDE 1.8.13** and **Arduino_Core_STM32 1.9.0**.

## Installation
Clone the repository to the Arduino libraries folder.

For example, on the Windows OS path is:
```
C:\Users\{User}\Documents\Arduino\libraries
```

For more information please visit: https://www.arduino.cc/en/guide/libraries#toc5

## Usage

This library is compatible with the **Arduino Servo library**, 
the only difference is that the class name is **ServoHT** instead of Servo.

For more information about the Arduino Servo library 
please visit: http://www.arduino.cc/en/Reference/Servo

## License
Copyright (C) 2021 PR-DC <info@pr-dc.com>

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as 
published by the Free Software Foundation, either version 3 of the 
License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.


CHANGES.txt
--------------------
PRDC_ServoHT
STM32 Servo Library using HardwareTimer from Arduino_Core_STM32
--------------------

Release 1.0.2, 25.01.2022.

- Arduino library specification

--------------------

Release 1.0.0, 27.02.2021.

- Initial release

--------------------


src/PRDC_ServoHT.h
/**
 * PRDC_ServoHT.h - Servo HardwareTimer based library
 * Author: Milos Petrasinovic <mpetrasinovic@pr-dc.com>
 * PR-DC, Republic of Serbia
 * info@pr-dc.com
 * 
 * --------------------
 * Copyright (C) 2021 PR-DC <info@pr-dc.com>
 *
 * This file is part of PRDC_ServoHT.
 *
 * PRDC_ServoHT is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as 
 * published by the Free Software Foundation, either version 3 of the 
 * License, or (at your option) any later version.
 * 
 * PRDC_ServoHT is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public License
 * along with PRDC_ServoHT.  If not, see <https://www.gnu.org/licenses/>.
 *
 */
 
#ifndef _PRDC_SERVOHT_H_
#define _PRDC_SERVOHT_H_
#include "Arduino.h"
#include <HardwareTimer.h>

#define Servo_VERSION           2     // software version of this library

#define MIN_PULSE_WIDTH       544     // the shortest pulse sent to a servo
#define MAX_PULSE_WIDTH      2400     // the longest pulse sent to a servo
#define DEFAULT_PULSE_WIDTH  1500     // default pulse width when servo is attached
#define REFRESH_INTERVAL    20000     // minumim time to refresh servos in microseconds

#define MAX_SERVOS  254

#define INVALID_SERVO 255

typedef struct  {
  uint8_t nbr;            // a pin number from 0 to 255
  uint8_t isActive;       // true if this channel is enabled, pin not pulsed if false
  uint8_t prevAttached;       // true if this channel is previusly attached
} ServoPin_t;

typedef struct {
  ServoPin_t Pin; 
  volatile unsigned int ticks;
  uint32_t channel;
  HardwareTimer *TIM;
} servo_t;

class ServoHT {
  public:
    ServoHT();
    uint8_t attach(int pin);           // attach the given pin to the next free channel, sets pinMode, returns channel number or 0 if failure
    uint8_t attach(int pin, int min, int max); // as above but also sets min and max values for writes.
    uint8_t attach();
    void detach();

    void write(int value);             // if value is < 200 its treated as an angle, otherwise as pulse width in microseconds
    void writeMicroseconds(int value); // Write pulse width in microseconds
    int read();                        // returns current pulse width as an angle between 0 and 180 degrees
    int readMicroseconds();            // returns current pulse width in microseconds for this servo (was read_us() in first release)
    bool attached();                   // return true if this servo is attached, otherwise false
  private:
    uint8_t servoIndex;               // index into the channel data for this servo
    int8_t min;                       // minimum is this value times 4 added to MIN_PULSE_WIDTH
    int8_t max;                       // maximum is this value times 4 added to MAX_PULSE_WIDTH
};
#endif // _PRDC_SERVOHT_H_


PRDC_ServoHT.cpp
/**
 * PRDC_ServoHT.cpp - Servo HardwareTimer based library
 * Author: Milos Petrasinovic <mpetrasinovic@pr-dc.com>
 * PR-DC, Republic of Serbia
 * info@pr-dc.com
 * 
 * --------------------
 * Copyright (C) 2021 PR-DC <info@pr-dc.com>
 *
 * This file is part of PRDC_ServoHT.
 *
 * PRDC_ServoHT is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as 
 * published by the Free Software Foundation, either version 3 of the 
 * License, or (at your option) any later version.
 * 
 * PRDC_ServoHT is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public License
 * along with PRDC_ServoHT.  If not, see <https://www.gnu.org/licenses/>.
 *
 */
 
#include "PRDC_ServoHT.h"

static servo_t servos[MAX_SERVOS]; // static array of servo structures
uint8_t ServoCount = 0; // the total number of attached servos
uint16_t maxValue = 65535;

#define SERVO_MIN() (MIN_PULSE_WIDTH - this->min * 4)   // minimum value in uS for this servo
#define SERVO_MAX() (MAX_PULSE_WIDTH - this->max * 4)   // maximum value in uS for this servo

// ServoHT()
// Object constructor
// --------------------
ServoHT::ServoHT(){
  if (ServoCount < MAX_SERVOS) {
    this->servoIndex = ServoCount++; // assign a servo index to this instance
    servos[this->servoIndex].ticks = DEFAULT_PULSE_WIDTH;   // store default values
  } else {
    this->servoIndex = INVALID_SERVO;  // too many servos
  }
}

// attach()
// Attach servo object to pin
// --------------------
uint8_t ServoHT::attach(int pin){
  return this->attach(pin, MIN_PULSE_WIDTH, MAX_PULSE_WIDTH);
}

uint8_t ServoHT::attach(){
  if(servos[this->servoIndex].Pin.prevAttached) {
    return this->attach(servos[this->servoIndex].Pin.nbr, this->min, this->max);
  } else {
    return false;
  }
}

uint8_t ServoHT::attach(int pin, int min, int max){
  if (this->servoIndex < MAX_SERVOS && !servos[this->servoIndex].Pin.prevAttached) {
    servos[this->servoIndex].Pin.nbr = pin;
    servos[this->servoIndex].ticks = DEFAULT_PULSE_WIDTH;
     
    TIM_TypeDef *Instance = (TIM_TypeDef *)pinmap_peripheral(digitalPinToPinName(pin), PinMap_PWM);
    uint32_t channel = STM_PIN_CHANNEL(pinmap_function(digitalPinToPinName(pin), PinMap_PWM));
    HardwareTimer *SERVO_TIMER = new HardwareTimer(Instance);
    servos[this->servoIndex].TIM = SERVO_TIMER;
    servos[this->servoIndex].channel = channel;
    
    // todo min/max check: abs(min - MIN_PULSE_WIDTH) /4 < 128
    this->min  = (MIN_PULSE_WIDTH - min) / 4; //resolution of min/max is 4 uS
    this->max  = (MAX_PULSE_WIDTH - max) / 4;
    servos[this->servoIndex].TIM->setMode(servos[this->servoIndex].channel, TIMER_OUTPUT_COMPARE_PWM1, servos[this->servoIndex].Pin.nbr);
    servos[this->servoIndex].TIM->setOverflow(REFRESH_INTERVAL, MICROSEC_FORMAT);
    writeMicroseconds(DEFAULT_PULSE_WIDTH);
    servos[this->servoIndex].TIM->resume();
    servos[this->servoIndex].Pin.isActive = true;
    servos[this->servoIndex].Pin.prevAttached = true;
  } else {
    servos[this->servoIndex].TIM->resumeChannel(servos[this->servoIndex].channel);
    servos[this->servoIndex].Pin.isActive = true;
  }
  return this->servoIndex;
}

// detach()
// Detach servo object from pin
// --------------------
void ServoHT::detach(){
  if(servos[this->servoIndex].Pin.prevAttached) {
    servos[this->servoIndex].TIM->pauseChannel(servos[this->servoIndex].channel);
    servos[this->servoIndex].Pin.isActive = false;
  }
}

// write()
// Write angle in degrees to servo
// --------------------
void ServoHT::write(int value){
  // treat values less than 544 as angles in degrees (valid values in microseconds are handled as microseconds)
  if (value < MIN_PULSE_WIDTH) {
    if (value < 0) {
      value = 0;
    } else if (value > 180) {
      value = 180;
    }

    value = map(value, 0, 180, SERVO_MIN(), SERVO_MAX());
  }
  writeMicroseconds(value);
}

// writeMicroseconds()
// Write angle in microseconds to servo
// --------------------
void ServoHT::writeMicroseconds(int value){
  // calculate and store the values for the given channel
  uint8_t index = this->servoIndex;
  if ((index < MAX_SERVOS)) {  // ensure channel is valid
    if (value < SERVO_MIN()) {        // ensure pulse width is valid
      value = SERVO_MIN();
    } else if (value > SERVO_MAX()) {
      value = SERVO_MAX();
    }

    servos[index].ticks = value;
    uint32_t compare_value = (float)maxValue/REFRESH_INTERVAL*value;
    servos[this->servoIndex].TIM->setCaptureCompare(servos[this->servoIndex].channel, compare_value, RESOLUTION_16B_COMPARE_FORMAT);
  }
}

// read()
// Read angle in degrees from servo
// --------------------
int ServoHT::read(){ // return the value as degrees
  return map(readMicroseconds() + 1, SERVO_MIN(), SERVO_MAX(), 0, 180);
}

// readMicroseconds()
// Read angle in microseconds from servo
// --------------------
int ServoHT::readMicroseconds(){
  unsigned int pulsewidth;
  if (this->servoIndex != INVALID_SERVO) {
    pulsewidth = servos[this->servoIndex].ticks;
  } else {
    pulsewidth  = 0;
  }

  return pulsewidth;
}

// attached()
// Check if servo is attached
// --------------------
bool ServoHT::attached(){
  return servos[this->servoIndex].Pin.isActive;
}
